[Fleet] Fix status summary when showUpgradeable is selected

criamico opened the pull request:
Fixes https://github.com/elastic/kibana/issues/159446

## Summary

Fix status summary in agents list. The problem is not actually in the UI but in the `GET api/fleet/agents` endpoint, specifically when querying with `getStatusSummary` and `showUpgradeable` (query used in the UI to show the status bar above the table:
```
/api/fleet/agents?getStatusSummary=true&perPage=5&showUpgradeable=true
```
This codepath was not taking into account the upgradeable agents so the results were wrong, both in the API and in the table header:
![Screenshot 2024-02-21 at 10 38 19](https://github.com/elastic/kibana/assets/16084106/d66a0e55-4e14-4874-8264-757eba9017eb)

### Testing
Enroll some agents and make sure that at least one of them is upgradeable (version < fleet server). Then select the `upgrade available` button on top and check that the values shown in the summary update correctly.

### Checklist

- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


<!--ONMERGE {"backportTargets":["8.13"]} ONMERGE-->

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

criamico commented:
/ci

criamico commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

criamico commented on x-pack/plugins/fleet/public/applications/fleet/sections/agents/agent_list_page/hooks/use_fetch_agents_data.tsx:
```diff
@@ -33,7 +33,7 @@ export function useFetchAgentsData() {
   const { displayAgentMetrics } = ExperimentalFeaturesService.get();
 
   const { notifications } = useStartServices();
-  // useBreadcrumbs('agent_list');
```
This was a leftover from my previous refactor

juliaElastic commented on x-pack/plugins/fleet/server/services/agents/crud.test.ts:
```diff
@@ -281,8 +288,136 @@ describe('Agents CRUD test', () => {
+      });
+
+      expect(result).toEqual({
+        agents: [
```
nit: is it necessary to verify the complete result? You can use `expect.objectContaining` to only include fields which are relevant to the test.

juliaElastic commented on x-pack/plugins/fleet/server/services/agents/crud.ts:
```diff
@@ -34,6 +34,17 @@ import { getLatestAvailableVersion } from './versions';
 
 const INACTIVE_AGENT_CONDITION = `status:inactive OR status:unenrolled`;
 const ACTIVE_AGENT_CONDITION = `NOT (${INACTIVE_AGENT_CONDITION})`;
+const AGENT_STATUS = [
```
I wonder if we could do a refactor to avoid repeating the statuses from here: https://github.com/elastic/kibana/blob/80c655912e5723a1cc4389aa4dd8263d5b43c657/x-pack/plugins/fleet/common/types/models/agent.ts#L20

Something similar was done here:
https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/common/constants/agent.ts#L29-L37
https://github.com/elastic/kibana/blob/80c655912e5723a1cc4389aa4dd8263d5b43c657/x-pack/plugins/fleet/common/types/models/agent.ts#L62-L63

criamico commented:
@juliaElastic Thanks for the review! I addressed the comment, but I have a doubt. Do you think that [this logic](https://github.com/elastic/kibana/pull/177618/files#diff-f7377ed9ad56eaa8ea188b64e957e771ccc7a7652fd1eaf44251c25b930f8448R356-R358) could cause slowness at scale or stop working altogether?

criamico commented:
@elasticmachine merge upstream

criamico commented:
@juliaElastic I used your suggestion but I kept the agent statuses refactor, I think it's still valuable plus it should be harmless.

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.13|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/177741">](https://github.com/elastic/kibana/pull/177741)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
